### DISCUSSION TITLE
[MTR05] 想問一下 week2作業挑戰題LIOJ1047

### DISCUSSION THREAD

**Discussion by cmtilo** [Lidemy/forum#69] [程式導師實驗計畫]
> 現在有一個排序好的陣列 arr，裡面的元素都是正整數而且保證不會重複。
給你一個數字 n，請寫出一個函式 search 回傳 n 在這個陣列裡面的 index，沒有的話請回傳 -1。

week2挑戰題我自己在sublime寫可以得到正確回傳值
不過題目後半段提到可以去[LIOJ 1047]...這邊就不行了一直submit都錯

> 寫完如果想確認是否正確，可以到這邊來測試，只是輸出入要改一下，然後要學一下 LIOJ 的使用方法：[LIOJ 1047](https://oj.lidemy.com/problem/1047)

以下是我目前的寫法..不曉得自己錯在哪裡...
```JS

function solve(lines) {
 var A = [];
 var N = parseInt(lines[0].split(" ",1));
 var M = parseInt(lines[0].slice(-1));
 var P = [];
 
 for(i=1; i>=N; i++){
 	A.push(lines[i])
 }

 for(i=N+1; i>=N+M; i++){
 	P.push(lines[i])
 }
 
 for(i=0; i<P.length; i++){
 	var n = P[i]
 	console.log(A.indexOf(n))
 }
 	
}
```

希望各位協助解惑...謝謝

**Comment by JAS0NHUANG** [comment]
如果你貼上來的 code 就是你在本機測試的 code 的話，那應該也不會輸出正確的值阿~~

首先關於這行：
>  var M = parseInt(lines[0].slice(-1));

你可以試試看如果 `lines[0]` 是 `5 12`，M 會是什麼？ 

還有就是兩個迴圈，它們真的有跑到嗎？:smile:

最後，你直接用 `indexOf()` 找數字的 index 並不能完成挑戰題比「一個一個搜尋」更快的要求，當然，還是可以先把你這一段程式修好，在 LIOJ 上 AC 了之後再去思考怎麼完成挑戰題的要求。

先給你一些方向，如果真的還是沒辦法解決我再給你更詳細的解答囉！

**Reply by cmtilo** [reply]
好的我先試試看這個方向...（自己測試時是直接用陣列測試）
而且我之前好像誤會題目意思了（以為是程式碼越短的意思）😳
我冷靜一下重新寫

**Comment by cmtilo** [comment]
我重新寫過目前本機測試結果正確了
助教的問題我也都改正了
但是現在submit還是WA
是否還是有沒注意到的錯誤
```JS
function solve(lines) {
	var str = lines[0].toString()
	var sep = str.split(" ")
	var N = parseInt(sep[0])//5
	var M = parseInt(sep[1])//3
	var A = [] //lines[1]~lines[N]
	var P = [] //lines[N+1]~lines[N+M]

	for(i=1; i<=N; i++){
		A.push(lines[i])
	}//1,2,3,4,5

	for(j=(N+1); j<=(N+M); j++){			
		P.push(lines[j])
	}//100,3,6		 
		

	function search(arr, n){
		var start = 0
		var end = arr.length-1 
		
		while (start <= end){	
			var mid = Math.floor((start + end) / 2) 
			if (n === arr[mid]){
				return mid
			} else if(n < arr[mid]){
				end = mid - 1
			} else if(n > arr[mid]){
				start = mid + 1
			}
		}
		return -1				
	}

	for(k=0; k<P.length; k++){
		console.log(search(A, P[k]))
	}

}
```

**Reply by JAS0NHUANG** [reply]
就一個問題：型態！:smile:

**Reply by cmtilo** [reply]
我有加上parseInt還是不行....
我有發現我用indexOf的話答案sumbit會是正確的，所以我又試著用typeof去看indexOf跟二分法return的輸出型態，結果卻都是number..
那這樣都是number為什麼不行...?

**Reply by JAS0NHUANG** [reply]
> 我有加上parseInt還是不行....

你可能要說清楚你是在哪裡加上 parseInt 的。

> 我有發現我用indexOf的話答案sumbit會是正確的，所以我又試著用typeof去看indexOf跟二分法return的輸出型態，結果卻都是number..

`indexOf` 的比較方式跟你自已寫一個 `search()` 函式不一樣。

> 那這樣都是number為什麼不行...?

哪些東西是 number 哪些不是呢？

可以用這組測資試試看喔 `solve(["4 2", "1", "3", "4", "22", "1", "22"]);`

**Reply by cwc329** [reply]
~~可以用這組測資試試看喔 `solve(["4 2", "1", "3", "4", "22", "1", "22"]);`
這組測資有符合題目規範嗎？~~
當我沒說((X

**Reply by cmtilo** [reply]
我在mid前面加上parseInt
然後去看mid型態
```JS
function search(arr, n){
var start = 0
var end = arr.length-1 

while (start <= end){	
	var mid = Math.floor((start + end) / 2) 
	console.log(typeof mid)	
	if (n === arr[mid]){
		return parseInt(mid)
	} else if(n < arr[mid]){
		end = mid - 1
	} else if(n > arr[mid]){
		start = mid + 1
	}
}
console.log(typeof mid)	
	

return -1

}
```

**Reply by JAS0NHUANG** [reply]
想一下喔：
1. mid  有需要加 `parseInt()` 嗎？:smile: 
2. 哪幾個東西是真的需要 `parseInt()` 的？（明確一點的提示你好了，你從 arr 裡面拿出來的東西是什麼型態呢？）

**Reply by cmtilo** [reply]
我成功了!!!!!!!我要流淚了🤣
謝謝助教們

**Comment by cmtilo** [comment]
還有一個小疑惑..為什麼mid變數設在while迴圈之前就不能執行了(好像變成無限迴圈了)
如下：
```JS
function search(arr, n){
	var start = 0
	var end = arr.length-1 
	var mid = Math.floor((start + end) / 2) 
	while (start <= end){			
		if (n === arr[mid]){
			return mid
		} else if(n < arr[mid]){
			end = mid - 1
		} else if(n > arr[mid]){
			start = mid + 1
		}
	}
	return -1				
}
```

**Reply by JAS0NHUANG** [reply]
想一下你每次跑 while 迴圈的時候 mid 值是多少？
什麼樣的狀況第一個 `if (n === arr[mid])` 會是 true？
